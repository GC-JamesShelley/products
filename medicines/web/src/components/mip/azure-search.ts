const azureApiVersion = process.env.AZURE_API_VERSION;
const azureIndex = process.env.AZURE_INDEX;
const azureKey = process.env.AZURE_KEY;
const azureService = process.env.AZURE_SERVICE;
const azureWordFuzziness = process.env.AZURE_WORD_FUZZINESS;

enum DocType {
  PilLabel,
  PilLabelAndLeaflet,
  PilLeaflet,
  Spc,
}

export interface IAzureSearchResult {
  '@search.score': number;
  '@search.highlights': { content: string[] };
  content: string;
  doc_type: DocType;
  metadata_storage_path: string;
}

const buildFuzzyQuery = (query: string): string =>
  query
    .split(' ')
    .map(word => `${word}~${azureWordFuzziness}`)
    .join(' ');

const buildAzureSearchUrl = (query: string): string => {
  const url = new URL(
    `https://${azureService}.search.windows.net/indexes/${azureIndex}/docs`,
  );

  url.searchParams.append('api-key', azureKey as string);
  url.searchParams.append('api-version', azureApiVersion as string);
  url.searchParams.append('highlight', 'content');
  url.searchParams.append('queryType', 'full');
  url.searchParams.append('search', query);

  return url.toString();
};

const getJson = async (url: string): Promise<any> => {
  const resp: Response = await fetch(url, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (resp.ok) {
    return resp.json();
  }
};

export const azureSearch = async (
  query: string,
): Promise<IAzureSearchResult[]> => {
  const body = await getJson(buildAzureSearchUrl(buildFuzzyQuery(query)));

  return body.value;
};
