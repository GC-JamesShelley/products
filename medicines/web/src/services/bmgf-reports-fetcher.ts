export const getReportList = async (
  containerUrl: string,
): Promise<string[]> => {
  return fetch(`${containerUrl}/upload-index.txt`, {
    method: 'GET',
    headers: {
      'Content-Type': 'text/text',
    },
  })
    .then((response) => response.text())
    .then((text) => {
      return text.split('\n').filter((report) => report.length > 0);
    });
};

const getReportUrl = async (
  reportToGet: string,
  containerUrl: string,
): Promise<any> => {
  return getReportList(containerUrl)
    .then((reports) => {
      return reports.find((report) => report.startsWith(reportToGet));
    })
    .then((report) => `${containerUrl}/${report}`);
};

export const getReportContent = async (
  report: string,
  containerUrl: string,
): Promise<string> => {
  return getReportUrl(report, containerUrl).then((reportUrl) => {
    return fetch(reportUrl, {
      method: 'GET',
      headers: {
        'Content-Type': 'text/html',
      },
    }).then((response) => response.text());
  });
};
