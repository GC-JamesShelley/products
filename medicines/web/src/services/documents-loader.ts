import DataLoader from 'dataloader';
import { IDocument } from '../model/substance';

const graphQlUrl = process.env.GRAPHQL_URL as string;

const fetchFromGraphQl = (query: string): Promise<Response> => {
  return fetch(graphQlUrl, {
    method: 'POST',
    mode: 'cors',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ query, variables: null }),
  });
};

const getDocumentsForProduct = async (productName: any) => {
  const query = `{ product(name: "${productName}") { documents {       productName
    activeSubstances
    title
    highlights
    created
    docType
    fileBytes
    name
    url } } }`;
  const response = await fetchFromGraphQl(query);

  return (await response.json()).data.product.documents;
};

export const documents = new DataLoader<string, IDocument[]>(
  async productNames => {
    return Promise.all(productNames.map(getDocumentsForProduct));
  },
);
