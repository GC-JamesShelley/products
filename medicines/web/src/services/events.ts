import TagManager from 'react-gtm-module';

const pushToDataLayer = (dataLayer: any) => {
  TagManager.dataLayer({
    dataLayer,
  });
};

export default {
  searchForProductsMatchingKeywords: (searchTerm: string, pageNo: number) => {
    pushToDataLayer({
      event: 'search',
      searchTerm,
      pageNo,
    });
  },
  viewProductsForSubstance: (substance: string) => {
    pushToDataLayer({ event: 'substance', substance });
  },
  viewSubstancesStartingWith: (letter: string) => {
    pushToDataLayer({ event: 'drugIndex', letter });
  },
  viewPage: (pageName: string) => {
    pushToDataLayer({ event: pageName });
  },
};
