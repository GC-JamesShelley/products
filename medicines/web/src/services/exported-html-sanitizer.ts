const updateImageTag = (imageNode, prefix) => {
  for (let i = 0; i < imageNode.attrs.length; i++) {
    if (imageNode.attrs[i].name === 'src') {
      const imageName = imageNode.attrs[i].value.split('/').pop();
      imageNode.attrs[i].value = encodeURI(`${prefix}${imageName}`);
    } else if (imageNode.attrs[i].name === 'v:shapes') {
      imageNode.attrs.splice(i, 1);
      i--;
    }
  }
  return imageNode;
};

const updateAnchorNameToId = (node) => {
  for (const attribute of node.attrs) {
    if (attribute.name === 'name') {
      attribute.name = 'id';
      return node;
    }
  }
  return node;
};

const removeUnwantedTableAttributes = (node) => {
  if (!node.attrs) {
    return node;
  }

  const unwantedAttributes = [
    'style',
    'v:shapes',
    'cellspacing',
    'cellpadding',
    'border',
    'width',
    'valign',
    'class',
  ];

  for (let i = 0; i < node.attrs.length; i++) {
    if (unwantedAttributes.includes(node.attrs[i].name)) {
      node.attrs.splice(i, 1);
      i--;
    }
  }

  return node;
};

const removeUnwantedAttributes = (node) => {
  if (!node.attrs) {
    return node;
  }

  const unwantedAttributes = ['style', 'align', 'class'];
  for (let i = 0; i < node.attrs.length; i++) {
    if (unwantedAttributes.includes(node.attrs[i].name)) {
      node.attrs.splice(i, 1);
      i--;
    }
  }
  return node;
};

const tagShouldBeRemoved = (tagName) => {
  return ['h1', 'o:p', 'w:sdt'].includes(tagName);
};

export const recurseNodes = (node, prefix) => {
  if (tagShouldBeRemoved(node.tagName)) {
    return;
  }
  if (node.tagName === 'img') {
    node = updateImageTag(node, prefix);
  } else if (node.tagName === 'td' || node.tagName === 'table') {
    node = removeUnwantedTableAttributes(node);
  } else if (node.tagName === 'a') {
    node = updateAnchorNameToId(node);
  } else {
    node = removeUnwantedAttributes(node);
  }

  for (let i = 0; i < node.childNodes?.length ?? 0; i++) {
    const returnedNode = recurseNodes(node.childNodes[i], prefix);
    if (returnedNode) {
      node.childNodes[i] = returnedNode;
    } else {
      node.childNodes.splice(i, 1);
      i--;
    }
  }
  return node;
};

export const cleanUpHtml = (htmlBody, assetPrefix) => {
  return recurseNodes(htmlBody, assetPrefix);
};

export const getHtmlBody = (htmlDoc) => {
  const html = htmlDoc.childNodes[0];
  for (const node of html.childNodes) {
    if (node.tagName === 'body') {
      return node;
    }
  }
};
