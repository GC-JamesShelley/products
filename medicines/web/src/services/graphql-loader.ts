import DataLoader from 'dataloader';
import { IProduct } from '../model/substance';

const graphQlUrl = process.env.GRAPHQL_URL as string;

const getProductsForSubstance = async (substanceName: string) => {
  const query = `
query ($substanceName: String) {
  substance(name: $substanceName) {
    products {
      name
      documentCount
    }
  }
}`;

  const variables = { substanceName };

  const response = await fetch(graphQlUrl, {
    method: 'POST',
    mode: 'cors',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query,
      variables,
    }),
  });

  return (await response.json()).data.substance.products;
};

const products = new DataLoader<string, IProduct[]>(async substanceNames => {
  return Promise.all(substanceNames.map(getProductsForSubstance));
});

export default {
  products,
};
