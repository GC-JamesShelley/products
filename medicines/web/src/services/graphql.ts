interface IGraphQLResponse<T> {
  data: T;
  // TODO: add proper type for errors field
}

const graphqlUrl = process.env.GRAPHQL_URL as string;

export async function graphqlRequest<T, V extends {}>({
  query,
  variables,
}: {
  query: string;
  variables: V;
}): Promise<IGraphQLResponse<T>> {
  const response = await fetch(graphqlUrl, {
    method: 'POST',
    mode: 'cors',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query,
      variables,
    }),
  });

  if (!response.ok) {
    throw new Error(`Error response code from GraphQL API: ${response.status}`);
  }

  return response.json();
}
