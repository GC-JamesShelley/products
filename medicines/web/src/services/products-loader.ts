import DataLoader from 'dataloader';
import { IProduct } from '../model/substance';
import { graphqlRequest } from './graphql';

const getProductsForSubstance = async (substanceName: string) => {
  const query = `
query ($substanceName: String) {
  substance(name: $substanceName) {
    products {
      name
      documentCount
    }
  }
}`;

  const variables = { substanceName };

  const resp = await graphqlRequest<
    { substance: { products: IProduct[] } },
    typeof variables
  >({ query, variables });

  return resp.data.substance.products;
};

export const products = new DataLoader<string, IProduct[]>(
  async substanceNames => {
    return Promise.all(substanceNames.map(getProductsForSubstance));
  },
);
