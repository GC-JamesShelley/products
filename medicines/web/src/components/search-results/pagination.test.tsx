import { shallow } from 'enzyme';
import React from 'react';
import { DocType } from '../../services/azure-search';
import Pagination from './pagination';

const page = 2;
const pageSize = 10;
const resultCount = 50;
const searchTerm = 'Caffeine';
const docTypes = [DocType.Pil, DocType.Par];

const dummyFunc = () => undefined;

describe(Pagination, () => {
  it('should render', () => {
    const component = shallow(
      <Pagination
        currentPage={page}
        pageSize={pageSize}
        resultCount={resultCount}
        searchTerm={searchTerm}
        enabledDocTypes={docTypes}
        handlePageChange={dummyFunc}
      />,
    );
    expect(component).toMatchSnapshot();
  });
});
