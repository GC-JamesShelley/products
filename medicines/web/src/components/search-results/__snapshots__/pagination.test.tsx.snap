// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render 1`] = `
<styled.nav>
  <ul
    className="pagination"
  >
    <li
      className="arrow"
      onClick={[Function]}
    >
      <a>
        Previous
      </a>
    </li>
    <div
      className="pagination-number"
    >
      <li>
        <a
          onClick={[Function]}
        >
          1
        </a>
        <span>
          –
        </span>
      </li>
      <li
        key="2"
      >
        2
        <span>
          –
        </span>
      </li>
      <li>
        <a
          onClick={[Function]}
        >
          3
        </a>
      </li>
      <div
        className="middle-group"
      />
      <li>
        …
      </li>
      <li>
        <a
          onClick={[Function]}
        >
          6
        </a>
      </li>
    </div>
    <li
      className="arrow"
      onClick={[Function]}
    >
      <a>
        Next
      </a>
    </li>
  </ul>
</styled.nav>
`;
